proposals/proposal_2020_12_rate-limiting.md

```diff
@@ -0,0 +1,200 @@
+- User set repo max in the setting page. (possibly set a default for repos at the org level)
+- When max builds is reached:
+  - queue the build (FIFO queue)
+  - builds triggered by [a configurable set of] events ignore queue
+
+pro:
+- protects system stability
+- gives users some control (which we'd be sure to get feedback for how they really want it to be)
+- expandable later down the line
+- allows prod deploys to be prioritized
+
+con:
+- doesn't prevent a queue from getting large for a team
+- lack of control around how builds are prioritized
+- a large team could run into issues with
+```
+
+#### Option A:
+```
+- System set max builds at the repo level.
+- Users can configure max builds in:
+  - settings page (upto to the system max)
+  - .vela.yml at the pipeline level (to enforce a single deployment or to avoid going to the settings page)
+  - if configured in both, .vela.yml takes priority
+- When max builds is reached:
+  - queue the build
+  - .vela.yml or settings page user set queue strategy [FIFO, priority queue, auto-cancel policy] default FIFO.
+
+pro:
+- protects system stability
+- maximum user configuration with a default to protect stability
+- default follows current user expectations
+- easy to enforce 1 deployment
+- can expand on the queue idea later
+
+con:
+- lots of different paths to test and implement.
+- not very MVP
+```
+
+
+#### Option B:
+```
+- System set max builds at the repo level.
+- Users can set a max builds in their settings repo settings page (up to the system max)
+- When max builds is reached, fail.
+
+Pro:
+- simple.
+- Easy to understand from a user perspective.
+
+Con:
+- to enforce 1 deployment, the user has to set to 1 before tagging
+- failing builds might make for bad UX
+  - likely lots of questions for build failures...
+```
+
+#### Option C:
+```
+same as B, but queue builds on failure.
+```
+
+#### Other options
+
+I could go on, but it would just be iterating on the different possibilities listed above.
+Do you see an option you'd prefer or tweak some?
+
+
+## Implementation
+
+<!-- Please leave this commented out section.
+
+This section is intended to explain how the solution will be implemented for the proposal.
+
+NOTE: If there are no current plans for implementation, please leave this section blank.
+-->
+
+**Please briefly answer the following questions:**
+
+1. Is this something you plan to implement yourself?
+2. What's the estimated time to completion?
+
+
+**Please provide all tasks (gists, issues, pull requests, etc.) completed to implement the design:**
+
+
+## Questions
+
+**Please list any questions you may have:**
+
+- I omitted time based limitation (e.g. N builds per hour) from my proposal as I didn't see a meaningful thing they provided that a concurrency based limitation didn't provide. But I'd be open to revisit this if there were ideas around it.
+- How could we utilize Redis's priority?
+- Would we have to setup a new queue?  
```

Comment:
Nice work and discussion @matt-fevold . I like the MVP option. Not to get too much into the work ahead but you would need much more than one more queue, right? Depending on what you mean by queue maybe?

Reply:
If this did end up involving a multi-queue design I think that would be fine. You could probably get away with a single queue and just manipulating the order within it. It's also important to remember queues are the main routing aspect of things like the `worker:` block in the yaml. Which means if new queues are dynamically added you do need to worry about namespace collisisions. 

An additional comment on a multi-queue design, if/when this evolves into that design the workers will also need to be properly compensated. Workers listen to a static set of queues when they are deployed. There might be a way to make it more dynamic but by the nature of how it works today, It's going to be very tricky if you want to dynamically add queues.

If you wanted to pick like 3-4 queues up front and then sort builds into those that's not hard at all because that 3-4 is a known static set that the worker can just subscribe to upon initialization.